File: backend/coinbaseScraper.py
import requests
from bs4 import BeautifulSoup
import re
import json
import os
from datetime import datetime, timedelta

def get_stored_data():
    try:
        if os.path.exists('coinbase_ranking.json'):
            with open('coinbase_ranking.json', 'r') as f:
                data = json.load(f)
                return data
        return None
    except Exception as e:
        print(f"Errore nella lettura del file: {e}")
        return None

def save_ranking(ranking):
    data = {
        'ranking': ranking,
        'timestamp': datetime.now().isoformat()
    }
    try:
        with open('coinbase_ranking.json', 'w') as f:
            json.dump(data, f)
    except Exception as e:
        print(f"Errore nel salvataggio del file: {e}")

def should_update():
    stored_data = get_stored_data()
    if not stored_data:
        return True
    
    last_update = datetime.fromisoformat(stored_data['timestamp'])
    return datetime.now() - last_update > timedelta(hours=24)

def fetch_ranking():
    try:
        url = "https://apps.apple.com/it/app/coinbase-compra-btc-e-cripto/id886427730"
        headers = {
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36',
            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8',
            'Accept-Language': 'it-IT,it;q=0.9,en-US;q=0.8,en;q=0.7'
        }
        
        response = requests.get(url, headers=headers, timeout=10)
        response.raise_for_status()
        
        soup = BeautifulSoup(response.text, 'html.parser')
        text = soup.get_text()
        match = re.search(r'#(\d+)\s+in\s+Finanza', text)
        
        if match:
            return int(match.group(1))
        return 0
            
    except Exception as e:
        print(f"Errore nello scraping: {e}")
        return 0

def get_app_ranking():
    try:
        stored_data = get_stored_data()
        
        if should_update():
            ranking = fetch_ranking()
            save_ranking(ranking)
            print(ranking)
        else:
            print(stored_data['ranking'])
            
    except Exception as e:
        print("0")

if __name__ == "__main__":
    get_app_ranking()

File: backend/api/coinbaseScraper.ts
import { NextApiRequest, NextApiResponse } from 'next';
import * as cheerio from 'cheerio';
import fs from 'fs';
import path from 'path';

const CACHE_FILE = path.join(process.cwd(), 'coinbase_ranking.json');

function getStoredData() {
  try {
    if (fs.existsSync(CACHE_FILE)) {
      const data = JSON.parse(fs.readFileSync(CACHE_FILE, 'utf-8'));
      return data;
    }
    return null;
  } catch (e) {
    console.log(`Errore nella lettura del file: ${e}`);
    return null;
  }
}

function saveRanking(ranking: number) {
  const data = {
    ranking: ranking,
    timestamp: new Date().toISOString()
  };
  try {
    fs.writeFileSync(CACHE_FILE, JSON.stringify(data));
  } catch (e) {
    console.log(`Errore nel salvataggio del file: ${e}`);
  }
}

async function shouldUpdate(force = false) {
  if (force) {
    return true;
  }

  const storedData = getStoredData();
  if (!storedData) {
    return true;
  }

  const lastUpdate = new Date(storedData.timestamp);
  const hoursSinceUpdate = (new Date().getTime() - lastUpdate.getTime()) / (1000 * 60 * 60);
  return hoursSinceUpdate > 24;
}

async function fetchRanking() {
  try {
    const url = "https://apps.apple.com/it/app/coinbase-compra-btc-e-cripto/id886427730";
    const response = await fetch(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36',
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8',
        'Accept-Language': 'it-IT,it;q=0.9,en-US;q=0.8,en;q=0.7'
      }
    });

    const html = await response.text();
    const $ = cheerio.load(html);
    const text = $.text();
    const match = text.match(/#(\d+)\s+in\s+Finanza/);

    if (match) {
      return parseInt(match[1]);
    }
    return 0;
  } catch (e) {
    console.log(`Errore nello scraping: ${e}`);
    return 0;
  }
}

export default async function handler(req: NextApiRequest, res: NextApiResponse) {
  try {
    const force = req.query.force === 'true';
    const storedData = getStoredData();

    if (await shouldUpdate(force)) {
      const ranking = await fetchRanking();
      saveRanking(ranking);
      console.log(ranking);
      return res.status(200).json(ranking);
    } else {
      console.log(storedData.ranking);
      return res.status(200).json(storedData.ranking);
    }
  } catch (e) {
    console.log("0");
    return res.status(200).json(0);
  }
} 